File: Algorithm/数组/有序数组的平方/sortedSquares.py
class Solution:
    def sortedSquares(self, A: list[int]) -> list[int]:
        return sorted([i**2 for i in A])

    def sortedSquares2(self, A: list[int]) -> list[int]:
        n = len(A)
        ans = [0] * n
        i, j, pos = 0, n - 1, n - 1
        while i <= j:
            if A[i] ** 2 > A[j] ** 2:
                ans[pos] = A[i] ** 2
                i += 1
            else:
                ans[pos] = A[j] ** 2
                j -= 1
            pos -= 1
        return ans

if __name__ == "__main__":
    print(Solution().sortedSquares([-4,-1,0,3,10]))
    print(Solution().sortedSquares2([-4,-1,0,3,10]))

File: Algorithm/leetcode/practice/数组/有序数组的平方/sortedSquares.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        vector<int> sortedSquares(vector<int>& A) {
            vector<int> res(A.size(), 0);
            int i = 0, j = A.size() - 1, pos = A.size() - 1;
            while (i <= j) {
                if (abs(A[i]) > abs(A[j])) {
                    res[pos] = A[i] * A[i];
                    i++;
                } else {
                    res[pos] = A[j] * A[j];
                    j--;
                }
                pos--;
            }
            return res;
        }
};

int main() {
    vector<int> A = {-4, -1, 0, 3, 10};
    Solution s;
    vector<int> res = s.sortedSquares(A);
    for (int i = 0; i < res.size(); i++) {
        cout << res[i] << endl;
    }
    return 0;
}